abdm_integrator/hip/views/care_contexts.py

```diff
@@ -0,0 +1,92 @@
+from abdm_integrator.hip.serializers.care_contexts import (
+    GatewayOnAddContextsSerializer,
+    LinkCareContextRequestSerializer,
+)
+from abdm_integrator.hip.views.base import HIPBaseView, HIPGatewayBaseView
+from abdm_integrator.utils import ABDMRequestHelper, cache_key_with_prefix, poll_for_data_in_cache
+
+
+class LinkCareContextRequest(HIPBaseView):
+
+    def post(self, request, format=None):
+        serializer = LinkCareContextRequestSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        self.check_for_already_linked(serializer.data)
+        gateway_request_id = self.gateway_add_care_contexts(serializer.data)
+        self.save_link_request(request.user, gateway_request_id, serializer.data)
+        response_data = poll_for_data_in_cache(cache_key_with_prefix(gateway_request_id))
+        return self.generate_response_from_callback(response_data)
+
+    def check_for_already_linked(self, request_data):
+        care_contexts_references = [care_context['referenceNumber']
+                                    for care_context in request_data['patient']['careContexts']]
+        linked_care_contexts = list(LinkCareContext.objects.
+                                    filter(care_context_number__in=care_contexts_references,
+                                           link_request__hip_id=request_data['hip_id'],
+                                           link_request__patient_reference=request_data['patient']
+                                           ['referenceNumber'],
+                                           link_request__status=LinkRequestStatus.SUCCESS
+                                           ).values_list('care_context_number', flat=True))
+        if linked_care_contexts:
+            code = HIPError.CODE_CARE_CONTEXT_ALREADY_LINKED
+            message = HIPError.CUSTOM_ERRORS[code].format(linked_care_contexts)
+            raise CustomError(
+                error_code=code,
+                error_message=message,
+            )
+
+    def gateway_add_care_contexts(self, request_data):
+        payload = ABDMRequestHelper.common_request_data()
+        payload['link'] = request_data
+        ABDMRequestHelper().gateway_post(HIPGatewayAPIPath.ADD_CARE_CONTEXTS, payload)
+        return payload['requestId']
+
+    @transaction.atomic()
+    def save_link_request(self, user, gateway_request_id, request_data):
+        link_request = LinkRequest.objects.create(user=user,
+                                                  gateway_request_id=gateway_request_id,
+                                                  hip_id=request_data['hip_id'],
+                                                  patient_reference=request_data['patient']['referenceNumber']
+                                                  )
+        for care_context in request_data['patient']['careContexts']:
+            LinkCareContext.objects.create(care_context_number=care_context['referenceNumber'],
```

Comment:
This should use [bulk_create](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#bulk-create) if possible

Reply:
Addressed in 98069dae7cba342724f71fa52aa46055a02a908d